internal/venice-common/src/main/java/com/linkedin/venice/writer/VeniceWriter.java

```diff
@@ -87,7 +88,7 @@ public class VeniceWriter<K, V, U> extends AbstractVeniceWriter<K, V, U> {
       new ChunkedPayloadAndManifest(null, null);
 
   // use for running async close and to fetch number of partitions with timeout from producer
-  private final ThreadPoolExecutor threadPoolExecutor;
+  private final Lazy<ThreadPoolExecutor> threadPoolExecutorLazy = Lazy.of(this::createThreadPoolExecutor);
```

Comment:
Should we make this static? Do we really want one thread pool per VW? We can have a lot of VW per server...

Reply:
But with single pool, you won't be able to limit the thread count per VW? Or is it still possible?

Reply:
The limit would need to be higher, but do we really need some per-VW threads? If we have 1K VWs in a server, do we really want 2K threads spread across 1K TPs? Would it be enough to have a single static TP with, e.g., max 100 threads, and other tasks get queued if we tap out all threads?

Reply:
Hmm, yeah that's probably more reasonable. 
I thought the 2 per VW is required for XInfra purpose. Maybe @sushantmane knows better about this...

Reply:
Originally I was thinking about having just one ThreadPool for all VWs.  But I wanted let callers's have some control over the concurrency of close operations which is where this thread pool is primarily used.

Reply:
>I thought the 2 per VW is required for XInfra purpose. Maybe @sushantmane knows better about this..
Threads here perform close operation (end all segments + close producer) which is blocking call.

Reply:
We could potentially make both endAllSegments and producer::close async and use static pool for VWs.

Reply:
Will do some experiments with the above changes.

Reply:
Maybe another way is to set the TP to have a `corePoolSize` of zero. Then with the `keepAliveTime` of 5 seconds, those threads will die out quickly after being used.

Reply:
Good point. Setting `this.threadPoolExecutor.allowCoreThreadTimeOut(true);` achieves the same effect which is done inside `VW::createThreadPoolExecutor` method.

Reply:
Ah I see... so I guess it could be fine then...

Reply:
@FelixGV  - Should we merge this PR then?

Reply:
I'm changing the scope of this PR slightly:
1) Keep Per ThreadPoolExecutor for now
2) Keep lazy init of PubSubProducer and ThreadPool
3) Close segments concurrently (VW::endAllSegments)

Do let me know if there are other things that I should address. Thanks.

Reply:
Update PR with the above changes.

Reply:
Thanks, I left another review.